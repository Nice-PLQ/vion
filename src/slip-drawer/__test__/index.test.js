import { mount } from '@vue/test-utils';
import SlipDrawer from '..';
import SlipDrawerGroup from '../../slip-drawer-group';
import { triggerDrag } from '../../../test';

beforeAll(() => {
  // mock DOM的offsetWidth
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return 100;
      },
    },
  });
});

afterAll(() => {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return parseFloat(window.getComputedStyle(this).width) || 0;
      },
    },
  });
});

describe('SlipDrawer', () => {
  test('滑动展开', async () => {
    const wrapper = mount(SlipDrawer, {
      props: {
        leftItems: [
          { text: '置顶' },
        ],
        rightItems: [
          { text: '标记为已读' },
          { text: '删除' },
        ]
      }
    });
    const el = wrapper.find('.v-slip-drawer__wrapper');
    triggerDrag(el, 30, 0);
    await el.trigger('transitionend');
    expect(wrapper.element).toMatchSnapshot();
    expect(el.element.style.transform).toMatch(/100px/);
    expect(wrapper.vm.opened).toBe(true);
    expect(wrapper.emitted().opened).toBeTruthy();
  });

  test('left-click事件触发', async () => {
    const wrapper = mount(SlipDrawer, {
      props: {
        leftItems: [
          { text: '置顶' },
        ],
      }
    });
    const el = wrapper.find('.v-slip-drawer__wrapper');
    triggerDrag(el, 300, 0);
    await el.trigger('transitionend');
    wrapper.find('.v-slip-drawer__action-btn').trigger('click');
    expect(wrapper.emitted()['left-click']).toBeTruthy();
  });

  test('right-click事件触发', async () => {
    const wrapper = mount(SlipDrawer, {
      props: {
        rightItems: [
          { text: '删除' },
        ]
      }
    });
    const el = wrapper.find('.v-slip-drawer__wrapper');
    triggerDrag(el, -300, 0);
    await el.trigger('transitionend');
    wrapper.find('.v-slip-drawer__action-btn').trigger('click');

    expect(wrapper.emitted()['right-click']).toBeTruthy();
  });

  test('小于offset时不展开', async () => {
    const wrapper = mount(SlipDrawer, {
      props: {
        offset: 50,
        rightItems: [
          { text: '删除' },
        ]
      }
    });
    const el = wrapper.find('.v-slip-drawer__wrapper');
    triggerDrag(el, -30, 0);
    await el.trigger('transitionend');
    expect(wrapper.vm.opened).toBe(false);
  });

  test('disabled禁用', async () => {
    const wrapper = mount(SlipDrawer, {
      props: {
        disabled: true,
        rightItems: [
          { text: '删除' },
        ]
      }
    });
    const el = wrapper.find('.v-slip-drawer__wrapper');
    triggerDrag(el, -30, 0);
    await el.trigger('transitionend');
    expect(wrapper.vm.opened).toBe(false);
  });

  test('组合', async () => {
    const wrapper = mount({
      template: `
        <v-slip-drawer-group>
          <v-slip-drawer :right-items="rightItems">test</v-slip-drawer>
          <v-slip-drawer :right-items="rightItems">test</v-slip-drawer>
        </v-slip-drawer-group>
      `,
      data() {
        return {
          disabled: true,
          rightItems: [
            { text: '删除' },
          ]
        };
      }
    }, {
      global: {
        components: { VSlipDrawerGroup: SlipDrawerGroup, VSlipDrawer: SlipDrawer }
      }
    });

    const el = wrapper.findAll('.v-slip-drawer__wrapper');
    triggerDrag(el[0], -30, 0);
    await el[1].trigger('transitionend');
    expect(el[1].element.style.transform).not.toMatch(/100px/);
    await wrapper.setProps({ disabled: false }); // 取消禁用
    triggerDrag(el[1], -30, 0);
    await el[1].trigger('transitionend');
    expect(wrapper.element).toMatchSnapshot();
    expect(el[1].element.style.transform).toMatch(/100px/);
  });
});
