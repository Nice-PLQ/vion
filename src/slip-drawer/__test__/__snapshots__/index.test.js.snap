// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlipDrawer 滑动展开 1`] = `
<div
  class="v-slip-drawer"
>
  <div
    class="v-slip-drawer__wrapper"
    style="transform: translate3d(100px, 0, 0); transition: transform .4s;"
  >
    <div
      class="v-slip-drawer__action v-slip-drawer__action-left"
    >
      
      
      <button
        class="v-slip-drawer__action-btn"
      >
        置顶
      </button>
      
      
    </div>
    <div
      class="v-slip-drawer__content"
    >
      
      
    </div>
    <div
      class="v-slip-drawer__action v-slip-drawer__action-right"
    >
      
      
      <button
        class="v-slip-drawer__action-btn"
      >
        标记为已读
      </button>
      <button
        class="v-slip-drawer__action-btn"
      >
        删除
      </button>
      
      
    </div>
  </div>
</div>
`;

exports[`SlipDrawer 组合 1`] = `
<div
  class="v-slip-drawer-group"
>
  
  <div
    class="v-slip-drawer"
  >
    <div
      class="v-slip-drawer__wrapper"
      style="transform: translate3d(0px, 0, 0); transition: transform .4s;"
    >
      <div
        class="v-slip-drawer__action v-slip-drawer__action-left"
      >
        
        
        
        
      </div>
      <div
        class="v-slip-drawer__content"
      >
        
        test
        
      </div>
      <div
        class="v-slip-drawer__action v-slip-drawer__action-right"
      >
        
        
        <button
          class="v-slip-drawer__action-btn"
        >
          删除
        </button>
        
        
      </div>
    </div>
  </div>
  <div
    class="v-slip-drawer"
  >
    <div
      class="v-slip-drawer__wrapper"
      style="transform: translate3d(-100px, 0, 0); transition: transform .4s;"
    >
      <div
        class="v-slip-drawer__action v-slip-drawer__action-left"
      >
        
        
        
        
      </div>
      <div
        class="v-slip-drawer__content"
      >
        
        test
        
      </div>
      <div
        class="v-slip-drawer__action v-slip-drawer__action-right"
      >
        
        
        <button
          class="v-slip-drawer__action-btn"
        >
          删除
        </button>
        
        
      </div>
    </div>
  </div>
  
</div>
`;
